P4Plugin/Source/P4SpecCommand.cpp

```diff
@@ -10,49 +12,106 @@ class P4SpecCommand : public P4Command
 	{
 		ClearStatus();
 		m_Root.clear();
-		
+
+		Conn().Log().Info() << args[0] << "::Run()" << Endl;
+		std::string fallback_error = std::string();
+		if (args.size() > 1)
+			fallback_error = args[1];
+
 		const std::string cmd = std::string("client -o ") + Quote(task.GetP4Client());
-		
+
 		if (!task.CommandRun(cmd, this))
 		{
-			std::string errorMessage = GetStatusMessage();			
-			//This error is handled here instead of in P4LoginCommand as it only shows up when running the first non-login command (currently always spec)
+			//MFA: This error is handled here instead of in P4LoginCommand as it only shows up when running the first non-login command (currently always spec)
 			if (StatusContains(GetStatus(), "login2"))
 			{
 				ClearStatus();
-				GetStatus().insert(VCSStatusItem(VCSSEV_Error, "Your chosen Perforce server has multi-factor authentication enabled. To log in please use a visual client such as p4v or the \"p4 login\" or \"p4 login2\" CLI commands"));
+				std::string exePath = std::string(
+#if defined(_MACOS)
+						"/Applications/HelixMFA.app/Contents/MacOS/HelixMFA"
+#endif
+#if defined(_WINDOWS)
+						"\"C:\\Program Files\\Perforce\\HelixMFA.exe\""
+#endif
+#if defined(_LINUX)
+						"helixmfa" //Just to check if by change someone created a symlink in /usr/local/bin
+#endif
+				);
+				Conn().VerboseLine(exePath);
+				try {
+					POpen proc = POpen(exePath + std::string(" ") + task.GetP4Port() + std::string(" ") + task.GetP4User());
+#if !defined(_WINDOWS)
+					std::string line = std::string();
+					bool helixmfa_found = false;
+					while (proc.ReadLine(line)) {
+						Conn().VerboseLine(std::string("HelixMFA: ") + std::string(line));
+						//If first line reported by HelixMFA does not contain Authenticating, then we didnt find it
+						if (!helixmfa_found && line.find("Authenticating") == std::string::npos) {
+							const std::string notfound_error = "HelixMFA Authenticator could not be found. Download and install it to continue. https://www.perforce.com/downloads/helix-mfa-authenticator\n"
+							"If this error keeps showing after installation, you can try running HelixMFA manually with the following parameters: Port=" + task.GetP4Port() + ". User=" + task.GetP4User();
+							GetStatus().insert(VCSStatusItem(VCSSEV_Error, notfound_error));
+							Conn().Log().Notice() << GetStatusMessage() << Endl;
+						}
+						else {
+							helixmfa_found = true;
+						}
+					}; // Wait until stdout is closed
+					if (!helixmfa_found) return false;
+					//We are not closing the handle because PerforcePlugin is terminated after reconnection and closes them all. Thus it would fail
+					//proc.~POpen();
```

Comment:
wrong, this destructor get called all the same on the next line when closing scope and proc is destroyed automatically (no garbage collector in C++)

Reply:
removed comments